### PR TITLE
Add logits processors to enable logit_bias in OpenAI server

### PR DESCRIPTION
**Pull request by zacharyblank**
This PR makes it possible to define custom logits processors to alter the probability of token generation based on user defined code. This allows for the vLLM OpenAI server to accept requests with `logit_bias` too.

The BiasLogitsProcessor is included specifically for the OpenAI server to handle requests with `logit_bias`.

### CODE REVIEW

**Comment by LorenzoBoccaccia** [comment]
Looking at the interface it seems we don't have access to the model state, would it be possible to get on at least the tokens generated so far passed in input? That'd open up the possibility of stateful warpers (i.e. jsonformer)

File: vllm/model_executor/layers/sampler.py [hunk 1]
```diff
@@ -142,6 +146,22 @@ def _get_output_tokens(input_metadata: InputMetadata) -> List[List[int]]:
...
+    logits: torch.Tensor,
+    output_tokens: List[List[int]]
+) -> torch.Tensor:
+    for _, seq_group in enumerate(input_metadata.seq_groups):
+        _, sampling_params = seq_group
+        logits_processors = sampling_params.logits_processors
+
+        if logits_processors is not None:
+            for logits_processor in logits_processors:
+                logits = logits_processor(logits, output_tokens)
```
Comment (noamgat) [inline vllm/model_executor/layers/sampler.py:160]: In this call, you send `output_tokens` to `logits_processor()`. However, in the `LogitsProcessor` interface, the `output_tokens` parameter does not exist:

```
def __call__(self, logits: torch.tensor) -> torch.tensor:
```

How does it work?

**Comment by simon-mo** [comment]
Thank you for your PR! Now #1469 is merged with logits_processors API, can you help rebase and use that instead? The `logit_bias` will still be very useful for vLLM users.

**Comment by Qubitium** [comment]
@zacharyblank Any update? This will be greate for advanced users. vllm/outlines is nice but nothing beats pure code based state machine when it comes to performance.

**Comment by esmeetu** [comment]
Thanks for your contribution. Close this PR as this feature has been supported in https://github.com/vllm-project/vllm/pull/3027.
